glean-core/src/metrics/url.rs

```diff
@@ -0,0 +1,261 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::error_recording::{record_error, ErrorType};
+use crate::metrics::Metric;
+use crate::metrics::MetricType;
+use crate::storage::StorageManager;
+use crate::CommonMetricData;
+use crate::Glean;
+
+// The maximum number of characters a URL Metric may have, before encoding.
+const MAX_URL_LENGTH: usize = 2048;
+
+/// A URL metric.
+///
+/// Record an Unicode string value a URL content.
+/// The URL is length-limited to `MAX_URL_LENGTH` bytes.
+#[derive(Clone, Debug)]
+pub struct UrlMetric {
+    meta: CommonMetricData,
+}
+
+impl MetricType for UrlMetric {
+    fn meta(&self) -> &CommonMetricData {
+        &self.meta
+    }
+
+    fn meta_mut(&mut self) -> &mut CommonMetricData {
+        &mut self.meta
+    }
+}
+
+// IMPORTANT:
+//
+// When changing this implementation, make sure all the operations are
+// also declared in the related trait in `../traits/`.
+impl UrlMetric {
+    /// Creates a new string metric.
+    pub fn new(meta: CommonMetricData) -> Self {
+        Self { meta }
+    }
+
+    fn is_valid_url_scheme(&self, value: String) -> bool {
+        let mut splits = value.split(':');
+        if let Some(scheme) = splits.next() {
+            return !scheme.is_empty()
+                && scheme.chars().next().unwrap().is_ascii_alphabetic() // Safe unwrap on a non empty string.
+                && scheme.chars().all(|c| c.is_ascii_alphanumeric() || ['+', '-', '.'].contains(&c));
```

Comment:
We _could_ be smart, take out the `scheme.chars()` iterator like:

```suggestion
			if scheme.is_empty() { return false; }
			let mut chars = scheme.chars();
			return chars.next().unwrap().is_ascii_alphabetic()
                && chars.all(|c| c.is_ascii_alphanumeric() || ['+', '-', '.'].contains(&c));
```

I think.
But ... micro-optimization really